utils/python/CIME/SystemTests/system_tests_common.py

```diff
@@ -48,70 +51,142 @@ def __init__(self, case, expected=None):
+                except:
+                    success = False
+
+                time_taken = time.time() - start_time
+                with self._test_status:
+                    self._test_status.set_status(phase_name, TEST_PASS_STATUS if success else TEST_FAIL_STATUS, comments=("Time=%d" % int(time_taken)))
+
+                if not success:
+                    break
 
-    def has_failed(self):
-        return self._runstatus == "FAIL"
+        return success
 
-    def pass_test(self):
-        if not self.has_failed():
-            self._runstatus = "PASS"
+    def build_phase(self, sharedlib_only=False, model_only=False):
+        """
+        This is the default build phase implementation, it just does an individual build.
+        This is the subclass' extension point if they need to define a custom build
+        phase.
 
-    def has_passed(self):
-        return self._runstatus == "PASS"
+        PLEASE THROW EXCEPTION OR RETURN FALSE ON FAIL
+        """
+        return self.build_indv(sharedlib_only=sharedlib_only, model_only=model_only)
 
-    def build(self, sharedlib_only=False, model_only=False):
-        build.case_build(self._caseroot, case=self._case,
-                         sharedlib_only=sharedlib_only, model_only=model_only)
+    def build_indv(self, sharedlib_only=False, model_only=False):
+        """
+        Perform an individual build
+        """
+        return build.case_build(self._caseroot, case=self._case,
+                                sharedlib_only=sharedlib_only, model_only=model_only)
 
     def clean_build(self, comps=None):
         build.clean(self._case, cleanlist=comps)
 
     def run(self):
-        return self._run()
-
-    def _set_active_case(self, case):
         """
-        Use for tests that have multiple cases
+        Do NOT override this method, this method is the framework that controls
+        the run phase. run_phase is the extension point that subclasses should use.
         """
-        self._case = case
-        self._caseroot = case.get_value("CASEROOT")
-
-    def _run(self, suffix="base", coupler_log_path=None, st_archive=False):
+        success = True
+        start_time = time.time()
         try:
-            success = case_run(self._case)
-            if success and st_archive:
-                success = case_st_archive(self._case)
+            expect(self._test_status.get_status(MODEL_BUILD_PHASE) == TEST_PASS_STATUS,
+                   "Model was not built!")
+            with self._test_status:
+                self._test_status.set_status(RUN_PHASE, TEST_PENDING_STATUS)
 
-            if success and self._coupler_log_indicates_run_complete(coupler_log_path):
-                self.pass_test()
-            else:
-                success = False
-                self.fail_test()
+            success = self.run_phase()
+            if not isinstance(success, bool):
+                logger.warning("build_phase did not return a bool, assuming PASS!")
```

Comment:
'build_phase' should say 'run_phase' here.

It seems better to me to assume a FAIL if the return value doesn't match expectation; is there a reason you're assuming PASS here? (Is this perhaps to handle tests that don't return anything, so you get a None? If so, it would help to document that reason and maybe handle it specially... I still feel that, in general, it's best to assume a FAIL rather than PASS if you see something unexpected.)